? gtk.window.cpp.patch
? gtk_scroll_lineup_linedown.patch
? gtk_scroll_lineup_linedown_2.patch
? contrib/samples/foldbar/foldpanelbar/mondrian.xpm
? src/gtk/window.cpp.new
Index: src/gtk/scrolbar.cpp
===================================================================
RCS file: /pack/cvsroots/wxwidgets/wxWidgets/src/gtk/scrolbar.cpp,v
retrieving revision 1.52
diff -b -u -2 -r1.52 scrolbar.cpp
--- src/gtk/scrolbar.cpp	2004/12/14 21:08:04	1.52
+++ src/gtk/scrolbar.cpp	2005/01/05 10:55:41
@@ -37,4 +37,5 @@
 extern bool   g_blockEventsOnDrag;
 extern bool   g_blockEventsOnScroll;
+static wxEventType g_currentUpDownEvent = wxEVT_NULL;
 
 static const float sensitivity = 0.02;
@@ -71,4 +72,12 @@
     win->GetEventHandler()->ProcessEvent( event );
 
+    // throw a LINEUP / LINEDOWN event if necessary
+    if (g_currentUpDownEvent != wxEVT_NULL)
+      {
+	wxScrollEvent event( g_currentUpDownEvent, win->GetId(), value, orient );
+	event.SetEventObject( win );
+	win->GetEventHandler()->ProcessEvent( event );
+      }
+
 /*
     wxCommandEvent cevent( wxEVT_COMMAND_SCROLLBAR_UPDATED, win->GetId() );
@@ -81,5 +90,4 @@
 // "button_press_event" from slider
 //-----------------------------------------------------------------------------
-
 static gint gtk_scrollbar_button_press_callback( GtkRange *widget,
                                                  GdkEventButton *gdk_event,
@@ -88,4 +96,29 @@
     if (g_isIdle) wxapp_install_idle_handler();
 
+    // check if a LINEUP/LINEDOWN event must be thrown
+    // I suppose here the size of scrollbar top/bottom buttons is 16px height
+    if (gdk_event->type == GDK_BUTTON_PRESS && gdk_event->button == 1)
+      {
+	int scroll_height, mouse_pos;
+
+	// get the mouse position when the click is done 
+	if (widget->orientation == GTK_ORIENTATION_VERTICAL)
+	  {
+	    scroll_height = GTK_WIDGET(widget)->allocation.height - 16;
+	    mouse_pos = (int)gdk_event->y;
+	  }
+	else
+	  {
+	    scroll_height = GTK_WIDGET(widget)->allocation.width - 16;
+	    mouse_pos = (int)gdk_event->x;
+	  }
+	
+	// compare mouse position to scrollbar height
+	if  (mouse_pos > scroll_height)
+	  g_currentUpDownEvent = wxEVT_SCROLL_LINEDOWN;
+	else if (mouse_pos < 16)
+	  g_currentUpDownEvent = wxEVT_SCROLL_LINEUP;
+      }
+
 //  g_blockEventsOnScroll = TRUE;  doesn't work in DialogEd
 
@@ -125,4 +158,7 @@
     win->m_isScrolling = FALSE;
 
+    // reset the LINEUP/LINEDOWN flag when the mouse button is released
+    g_currentUpDownEvent = wxEVT_NULL;
+
     return FALSE;
 }
@@ -165,10 +201,8 @@
                         (GtkSignalFunc) gtk_scrollbar_callback,
                         (gpointer) this );
-
     gtk_signal_connect( GTK_OBJECT(m_widget),
                         "button_press_event",
                         (GtkSignalFunc)gtk_scrollbar_button_press_callback,
                         (gpointer) this );
-
     gtk_signal_connect( GTK_OBJECT(m_widget),
                         "button_release_event",

 	  	 
